### PR TITLE
Imagecodecs

### PR DESCRIPTION
**Pull request by hmaarrfk**
Improvements to conda-forge:
* Current package `imagecodecs-lite` is defunct and on pypi, is a pointer to `imagecodecs` built with no options.
* This package works with python to open compressed tiff images. Something currently **impossible** with `imageio` and `scikit-image` being installed with conda-forge.
* This package includes as many options in `imagecodecs` as conda-forge currently has.
* This package uses `jpeg9` instead of `jpeg-turbo` to be consistent with the rest of the conda-forge ecosystem.
* Worked with upstream to include options to overwrite their previously hardcoded options to options that would be amenable to conda forge.
   * https://github.com/cgohlke/imagecodecs/pull/2

Previous checklist for work that

TODO:
* [x] Try without custom build script for windows.
* [x] Try withoout JPEG turbo

Needs:
* [x] https://github.com/conda-forge/staged-recipes/pull/10327
* [x] https://github.com/conda-forge/lcms2-feedstock/pull/3 or https://github.com/conda-forge/lcms2-feedstock/pull/4 lcms2 for windows
* [x] https://github.com/conda-forge/libaec-feedstock/pull/5 libaec for windows
* [x] *needs* jpeg-turbo, will not find enum constants like `JCS_EXT_RGBA` in libjpeg
     * Basically i just disabled any jpeg. Seems to be OK. Will have to see what the user experience is like
* [x] libzopfli https://github.com/conda-forge/staged-recipes/pull/10513

May benefit from: 
* [x] https://github.com/conda-forge/staged-recipes/pull/10330
* [ ] https://github.com/conda-forge/staged-recipes/pull/10328  -- its apparently not a good idea to have libjpeg-turbo12 and libjpeg-turbo on the same program.

Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by Czaki** [comment]
I would like to share my experience with building imagecodecs wheels. I got the point with support of @cgohlke so setup.py is modified to simplify maintaining. 

Current setup.py detects that there is non windows build with all dependencies if `LD_LIBRARY_PATH` or `LIBRARY_PATH` environment variable is defined

I decide to not install libraries globally, this allow to try build on own machine without need of cleaning.

In my repossitotory there are few scripts:

1. To build all depends libraires I need to install few things. 
   1. On ubuntu: `nasm autopoint libhdf5-dev` 
   2. On macos (from `brew`): `pcre openssl nasm automake libtool`. Also default compiler does not support `openmp`. I found that it is enough to install `gcc@9` and use it as compiler.
2. `build_utils/download_libraries.sh` - download libraries in proper version. Downloaded libraries land in `build_utils/libs_src` directory.
3.  `build_utils/build_libraries.sh` - it builds all needed libraries. Builded libraries land in `build_utils/libs_build` directory. During build there are few patches apply (to allow build on macos). patches are in `build_utils/*.patch`)
Macos needs to  set path to openssl with
```
export LDFLAGS=-L/usr/local/opt/openssl/lib
export CPPFLAGS=-I/usr/local/opt/openssl/include
```
4. Because macos introduce SIP (System Integrity Protection) and then using of `LD_LIBRAY_PATH` is not possible then I upgrade loading path with `build_utils/fix_macos_lib.sh`

I also create docker images (base on manylinux2010). Their build recipes are in `build_utils`. They are also available on https://hub.docker.com/search?q=bokota&type=image (for speedup). Maybe they will be useful for theses builds? 

the upgrade to 2019.11.28 can be seen here: Czaki/imagecodecs/pull/2 (i'm waiting on tests result).

**Comment by Czaki** [comment]
State is accepted by @cgohlke and merged to master. Wheels created in this pipeline are uploaded to pypi https://pypi.org/project/imagecodecs/#files

**Comment by hmaarrfk** [comment]
nice, good update. Excited to try building again for a new release!

**Comment by Czaki** [comment]
Next release is ready.

**Comment by hmaarrfk** [comment]
don't we need jrxlib?
 nobody merged that in yet. so i'm kinda stuck.

**Comment by Czaki** [comment]
I think that best start case is to get version without additional extension (lines from 149 in setup.py). 
Then maybe add optional extensions.

**Comment by hmaarrfk** [comment]
hmm ok, i've been able to use imagecodecs-lite for now. Not sure I have time to get a working recepie for this immediately. Wanna give it a stab?

**Comment by Czaki** [comment]
I do not use anaconda currently. I work with classical CPython and install everything from pypi.

So I create pipeline to build wheel. Then I can share knowledge what I need to build wheel. 

I do not know how such package should be introduced to anaconda. how many recipes it should be created and otthers. 

Christopher really fast upgrade dependencies. There is no option for create recipe based on wheels?

**Comment by hmaarrfk** [comment]
> Christopher really fast upgrade dependencies

Do you mean that Christopher changes the dependencies often?

I'll see what I can do over the weekend.

**Comment by Czaki** [comment]
I see that he uses newest version of libraries which are available. But also it is possible that code may work with older versions of libraries but it is not checked.

**Comment by hmaarrfk** [comment]
Conda forge has REALLY updated versions of libraries.

Typically, library versions are not a problem.

**Comment by Czaki** [comment]
I think that I misunderstand your question. 
_imagecodecs module depends on jxrlib 
In my configuration this lib is represent by `'jxrglue', 'jpegxr'` in `libraries` variable.

**Comment by hmaarrfk** [comment]
hmm, cool, the package just got merged, so we can try again later.

**Comment by hmaarrfk** [comment]
how do I disable jxrlib?

**Comment by hmaarrfk** [comment]
It seems the problem is that the recipe has been regenerated after merged in staged recipe.

https://github.com/conda-forge/jxrlib-feedstock

it builds locally with jxrlib

**Comment by npavlovikj** [comment]
Hmm, the jxrlib-feedstock is empty?

**Comment by npavlovikj** [comment]
Looks like there is an issue with staged-recipes, and all packages merged today are not uploaded nor have feedstocks created, including jxrlib. We need to wait until the core team fixes this issue before restarting this build.

**Comment by hmaarrfk** [comment]
xref: https://github.com/conda-forge/staged-recipes/issues/10438

**Comment by hmaarrfk** [comment]
Needs: https://github.com/conda-forge/staged-recipes/pull/10513

**Comment by hmaarrfk** [comment]
@Czaki @cgohlke is there any way to disable certain libraries from being used? It seems that the `_imagecodecs.pyx` file tries to include all the plugins irrespectively.

**Comment by cgohlke** [comment]
> is there any way to disable certain libraries from being used? It seems that the _imagecodecs.pyx file tries to include all the plugins irrespectively.

Only codecs that cannot be built using current Debian packages are optional and disabled by default: jpeg12, jpegls, jpegxl, and zfp.

**Comment by hmaarrfk** [comment]
Thanks for the reply and clarification.

Do take this as a kind request to attempt to make new libraries optional if the packagers decide to explicitly disable them. I know it can be more maintenance work, but keeping up with the wide array of formats you support is challenging!

**Comment by Czaki** [comment]
From my side, on MacOS I meet conflict between zopfli library and zopflipy package (when testing) when both are compiled with gcc/g++ from brew. 
I use g++ to have OpenMP support for build but i need to compile zopflipy with clang.

**Comment by hmaarrfk** [comment]
I have a PR for zopfli in flight for conda forge I think

**Comment by hmaarrfk** [comment]
how did you install zopflipy? there is no feedstock for it?

I mean, conda-forge attempts to have binary compatibility within itself, not really beyond it.

**Comment by Czaki** [comment]
`zopflipy` has embedded full source inside. So I install it with `pip install zopflipy` 

The problem which I meet is that macos system `clang` does not support OMP. But when I globally change compiler on g++ (for OMP) and try to import both `imagecodecs` and `zopflipy` I got SIGABRT. 

The solution (for my case) is to use system compiler when install `zopflipy` eg.
`CC=clang CXX=clang pip install zopflipy`

**Comment by hmaarrfk** [comment]
ok, lets get to that issue once zopfli gets merged in.

**Comment by jni** [comment]
@hmaarrfk do you know whether any of the blockers on this recipe have been removed since you last looked at it?

**Comment by hmaarrfk** [comment]
I believe one of them was jpeg turbo. And that was probably the hardest one to solve

**Comment by Czaki** [comment]
> I believe one of them was jpeg turbo. And that was probably the hardest one to solve

Problem wit version or with compilation?

**Comment by hmaarrfk** [comment]
i just don't think people at conda-forge want turbo and non-turbo installed at the same time.

Therefore, we need to get all of conda-forge migrated...

That is the challenge.

**Comment by jni** [comment]
That doesn't sound like an easy challenge. ☹️

**Comment by hmaarrfk** [comment]
No, i think i'll have to patch out libjpeg from imagecodecs and see what happens.

Maybe this weekend?

**Comment by hmaarrfk** [comment]
@Czaki i'm having trouble adding the openjpeg library. Can you help with compilation options for windows?

**Comment by hmaarrfk** [comment]
I guess Charls is now the challenge.

**Comment by Czaki** [comment]
I only create wheels for linux and macos. Wheels for windows are made by @cgohlke. SO I do not have experience with windows compilation...

**Comment by cgohlke** [comment]
Re openjpeg on Windows: it looks like `set CFLAGS="%CFLAGS% /I%LIBRARY_INC%\openjpeg-%openjpeg%"` is not picked up by pip/setuptools/distutils. 
Can you try adding something like `EXTENSIONS['jpeg2k']['include_dirs'] = [LIBRARY_INC+ 'openjpeg-2.3']` to setup.py?

**Comment by hmaarrfk** [comment]
@csachs this is almost ready again. Are you still willing to be maintainer?

@jakirkham, @sdvillal i would appreciate a review in the context of the conda-forge ecosystem.

In particular, I omitted all jpeg turbo or non turbo linking.

**Comment by csachs** [comment]
> @csachs this is almost ready again. Are you still willing to be maintainer?
Yes, would still be fine to me. Thanks a lot for working on this all the time along!

**Comment by jakirkham** [comment]
Would it be possible to upstream the `setup.py` (or at least get that process going)? It seems that would be difficult to maintain in the recipe long term.

**Comment by hmaarrfk** [comment]
@jakirkham honestly, I'm not sure it is something upstream will be willing to solve, unless we convince them to use conda-pack (or i forget what the name of the technology was)

The linux configuration was quite painless, but the Windows configuration was plagued by the standard:

* Windows installations change how they decide to package the `.lib` files. Is it `png.lib`, or `libpng.lib`???

Most of the changes revolved around the correct naming of the conda-forge libraries, and disabling libraries which we don't have yet (and jpeg/turbo)

**Comment by Czaki** [comment]
When I create wheels for linux and macos my changes are introduced to setup.py, so I suggest to ask.

File: recipes/imagecodecs/setup.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    OPENMP_ARGS = ['/openmp']
+    bz2_libraries = ['bzip2']
+    jpeg2k_include_dirs = [
+        (os.environ.get('LIBRARY_INC', '') +
+         '\\openjpeg-' + os.environ.get('openjpeg', '2.3'))
+    ]
+    jpegls_libraries = ['charls-2-x64']
+    lz4_libraries = ['liblz4']
+    lzma_libraries = ['liblzma']
+    png_libraries = ['libpng', 'libz']
```
Comment (hmaarrfk) [inline recipes/imagecodecs/setup.py:82]:
```suggestion
    png_libraries = ['libpng', 'z']
```

**Comment by jakirkham** [comment]
Sure figured the `setup.py` was here because of complex issues. Same thing happens with custom `Makefile`s in recipes. Just saying it would be preferable from a maintenance standpoint, if upstream handled these build components. It seems in scope.

Just as some thoughts, it may be reasonable to ask them for environment variables to override library names or possible flags to `setup.py`. This is how some other libraries solve these problems.

**Comment by cgohlke** [comment]
> Would it be possible to upstream the setup.py

IIUC correctly, the additions to upstream setup.py to support conda-forge would be limited (so far):

```Python
elif os.environ.get('LIBRARY_INC', ''):
    # conda-forge

    del EXTENSIONS['jpeg8']
    del EXTENSIONS['jpeg12']
    del EXTENSIONS['jpegxl']
    del EXTENSIONS['jpegxr']
    del EXTENSIONS['zfp']

    if sys.platform == 'win32':

        EXTENSIONS['bz2']['libraries'] = ['bzip2']
        EXTENSIONS['jpeg2k']['include_dirs'] = [
            os.environ.get('LIBRARY_INC', '') +
            '\\openjpeg-' +
            os.environ.get('openjpeg', '2.3')
        ]
        EXTENSIONS['jpegls']['libraries'] = ['charls-2-x64']
        EXTENSIONS['lz4']['libraries'] = ['liblz4']
        EXTENSIONS['png']['libraries'] = ['libpng', 'z']
        EXTENSIONS['lzma']['libraries'] = ['liblzma']
        EXTENSIONS['webp']['libraries'] = ['libwebp']

```

It would be nice to support `JPEG8` and `JPEGXR` codecs. They are commonly used in TIFF and CZI.

Unfortunately I have not found a clean way to switch between `libjpeg` and `libjpeg-turbo` in setup.py, but decided to add a Cython DEF variable that can be changed in the `_jpeg8.pyx` file to switch between those libraries.

**Comment by hmaarrfk** [comment]
@jakirkham thanks for pushing and striving for better!

@cgohlke that seems like a good summary of what was found so far.

I'm not sure if that `elif` condition is correct.

Maybe @jakirkham's suggestion of environmetn variables is better, makign this more suitable for other downstream packagers as well in the future.

**Comment by hmaarrfk** [comment]
@cgohlke regarding jpeg libraries, is it meerly some defined constants that are different, or are you using some features explicit to turbo?

My understanding was that they were quite compatible (modulo the 12 bit support)

**Comment by cgohlke** [comment]
Compared to libjpeg v8, libjpeg_turbo supports some additional colorspaces, including alpha channel, which I would like to keep if possible. The Cython DEF macro works, but there seems to be no way to pass a value as an argument to setuptools.Extension().
Both libraries can be compiled for 12-bit after some patching...

**Comment by hmaarrfk** [comment]
Hmm, this makes a strong case to furthering the migration to jpeg 8 so we can finally switch to turbo.

As for your patch, conda-forge sets `LIBRARY_INC` to something non-zero.

**Comment by hmaarrfk** [comment]
@cgohlke do you have an official page where we can continue to discuss this?

**Comment by jakirkham** [comment]
I’d suggest we table the turbo discussion and focus on getting this through. There’s a separate thread about how to enable turbo in conda-forge and it’s somewhat hairy. Would like to make sure we don’t get blocked on that if that’s alright.

**Comment by hmaarrfk** [comment]
@jakirkham i'm totally fine tabling the turbo discussion. Happy to just move this forward.

Openning compressed tiffs has been a real pain for me !

**Comment by sdvillal** [comment]
Back in the day I added this turbo vs jpeg test to a recipe:
https://github.com/loopbio/opencv-feedstock/blob/loopbio/recipe/ensure_jpegturbo_opencv_plays_nicely_with_jpeg9.py

To earn some peace of mind, maybe we want something similar here?

**Comment by hmaarrfk** [comment]
@sdvillal is this something you are OK working through in the feedstock as opposed this recipe itself.

As is, i am not compiling with turbo or jpeg (non turbo)

image

**Comment by cgohlke** [comment]
I found a way to use Cython DEF macros in `setup.py` by overriding `setuptools.command.build_ext` so it should be possible to build against libjpeg >= v8 in the next version using a setup.py switch.

What is the problem with the `jpegxr` extension? AFAICT all the requirements are met by the conda-forge jxrlib 1.1 package.

> do you have an official page where we can continue to discuss this?

you can open an issue at https://github.com/cgohlke/imagecodecs

**Comment by jakirkham** [comment]
FWIW I'm happy to just merge and let anything else get resolved in the feedstock. It seems like it's taken a lot of work to get to this point. So am interested in making sure we don't get blocked here. Though want to make sure others are comfortable with that decision. Thoughts? 🙂

**Comment by hmaarrfk** [comment]
@jakirkham the only thing i'm worried about is that `jpeg` might be a more important dependency that I think it is.

However, I think recently imagecodecs-lite has been merged into imagecodecs so we should be no worse than `imagecodecs-lite` in terms of support.

Is that correct @cgohlke ?

If so, I would like to merge and fixup the rest in the feedstock.

**Comment by cgohlke** [comment]
> I think recently imagecodecs-lite has been merged into imagecodecs so we should be no worse than imagecodecs-lite in terms of support.

correct.

**Comment by jakirkham** [comment]
Am ok holding off if we are still trying to figure out `jpeg`. Agree that makes sense to get right.

**Comment by hmaarrfk** [comment]
Sorry @jakirkham 

The current package:
* imagecodecs-lite is now deprecated upstream and is simply a dummy package containing an empty setup.py file <- this is conda-forge's current state.
* This package, *does not* link to any jpeg libraries. For the cases where it does need jpeg, it falls back to the functionality that we used to have with `imagecodecs-lite`


As such, as is, this package is an upgrade on all fronts, meaning that we should totally move forward with it.

**Comment by sdvillal** [comment]
@hmaarrfk As you likely have seen I have PRed your PR with a jpeg read test. For me it fails when building the package locally. Is dropping jpeg decoding support intended?

Actually I do not quite understand what how libjpeg is used in this recipe. Since libjpeg-turbo is not a build dependency, I guess we are not building with jpeg support at all. And then, why do we need it as a runtime dependency?

If we do not fix the turbo problem in CF, I think the best solution would be to build against libjpeg 9 when that is supported in the next imagecodecs version.

I also think that jpegxr support is quite important.

**Comment by hmaarrfk** [comment]
Neither jpeg, nor jpeg turobo is used in this recipe.

This is intended as a first pass so as not to delay this critical package for the image analysis community.

Building with jpeg 9 is not supported upstream

**Comment by cgohlke** [comment]
imagecodecs-2020.5.30 can be built with libjpeg >= 8 instead of libjpeg-turbo.

**Comment by sdvillal** [comment]
Thanks @cgohlke, that should make it possible to build against libjpeg.

I have given it a try here: https://github.com/hmaarrfk/staged-recipes/pull/1. But it seems that it is still trying to compile against turbo (at least when I run that recipe locally) and so we are getting some errors.

Maybe the check for conda-forge should be based on the [value of the "CONDA_BUILD" variable ](https://docs.conda.io/projects/conda-build/en/latest/user-guide/environment-variables.html#environment-variables-set-during-the-build-process)?

**Comment by cgohlke** [comment]
The cf detection is based on https://github.com/conda-forge/staged-recipes/pull/10331#issuecomment-635669521: "conda-forge sets LIBRARY_INC to something non-zero."

If that does not work for you, provide your own `imagecodecs_distributor_setup.customize_build` module.function. See https://github.com/cgohlke/imagecodecs/blob/f38d0d4e5db1d08d92742a5d899325fc7c7b56e7/setup.py#L268-L314

Cython needs to be installed to re-cythonize the pyx code for libjpeg. The sdist ships with C code for libjpeg-turbo.

**Comment by sdvillal** [comment]
I see. As per [documentation](https://docs.conda.io/projects/conda-build/en/latest/user-guide/environment-variables.html#environment-variables-set-during-the-build-process) I think ```LIBRARY_INC``` is actually only set on Windows. We can easily workaround it by manually defining it in linux too, but for the next release of imagecodecs we should better test for ```CONDA_BUILD``` instead.

Using the ```LIBRARY_INC``` workaround allows the jpeg extension to be built against jpeg 9, that is great! We still have some problems with libjpegxr that I'm trying to troubleshoot.

**Comment by sdvillal** [comment]
@cgohlke After applying [this patch](https://github.com/hmaarrfk/staged-recipes/pull/1/commits/bcc977c600eb14225be5bfb729be170e2260cbc4) to setup.py, the conda package builds successfully, at least on linux. Is there an upstream repo for imagecodecs where I could submit these changes? Otherwise feel free to apply them for the next imagecodecs release.

@hmaarrfk if you merge the mentioned PR we should be able to finish this soon.

**Comment by hmaarrfk** [comment]
https://github.com/cgohlke/imagecodecs

**Comment by hmaarrfk** [comment]
@sdvillal thanks for your help and insight. do you want to be a maintainer as well?

**Comment by sdvillal** [comment]
I guess the jpeg tests were harmless, an assurance that we could read jpeg and something to keep in mind if we try to move this recipe to turbo without the whole stack doing so - but that should not happen. It is very fine to remove them.

It seems that the line endings of the patch have been converted from windows to linux at some point?

I'm ok with becoming a maintainer too.

**Comment by sdvillal** [comment]
@hmaarrfk I have now a PR open against imagecodecs:
https://github.com/cgohlke/imagecodecs/pull/2

If you want, I can open more PRs against your branch with the updated diff to try to fix this build. Do you see anything missing?

**Comment by hmaarrfk** [comment]
I'mjust going to give you access to my conda-forge repo, you can push directly.

**Comment by hmaarrfk** [comment]
I guess I just don't know how to deal with patches and line endings.

**Comment by sdvillal** [comment]
Actually I very much like the practical approach you have taken to just have a custom ```setup.py``` instead of the patch, I do not want to deal with these pesky line endings either. We can just remove the workaround when imagecodecs sees a new release with the patch.

**Comment by hmaarrfk** [comment]
It seems that you have a slightly different organization.

I would prefer that we tested your patch since you already submitted it.

Lets wait 10 mins until this build passes, then test your take.

**Comment by hmaarrfk** [comment]
I finally got my version working. testing now with yours. I think it is the same, but you know, there are always bugs and errors.

**Comment by cgohlke** [comment]
Not sure why you are still patching/overriding imagecodecs's setup.py instead of providing your own `customize_build` function in an importable `imagecodecs_distributor_setup` module?

**Comment by hmaarrfk** [comment]
That would hvae been my preferred solution, but you decided to graciously include our options upstream.

We wish to work with you to get our changes included in your future distributions so that we stay in synch. 

Thats all.

Your choice of line endings is totally fine, I was just venting my frustration at tools, not at your choice of them!

**Comment by hmaarrfk** [comment]
@jakirkham outstanding issues with jpeg have been resolved!

A merge would be greatly appreciated!

**Comment by cgohlke** [comment]
Does the testsuite pass? E.g. install imagecodecs and it's [test requirements](https://github.com/cgohlke/imagecodecs/blob/f38d0d4e5db1d08d92742a5d899325fc7c7b56e7/setup.py#L416-L417) (most are optional) and run `test_imagecodecs.py` from within the `tests` directory included in the sdist.

**Comment by hmaarrfk** [comment]
I'm trying locally now before pushing.

@cgohlke do you have an interest in being a maintainer. I think it would simply make it easier to ping you if we continue to have issues with future updates.

**Comment by hmaarrfk** [comment]
Seems like many tests are running, but some seem to be getting skipped

```
test_imagecodecs.py::test_tifffile[deflate-u1] SKIPPED                   [ 99%]
test_imagecodecs.py::test_tifffile[deflate-u2] SKIPPED                   [ 99%]
test_imagecodecs.py::test_tifffile[deflate-f4] SKIPPED                   [ 99%]
test_imagecodecs.py::test_tifffile[lzma-u1] SKIPPED                      [ 99%]
test_imagecodecs.py::test_tifffile[lzma-u2] SKIPPED                      [ 99%]
test_imagecodecs.py::test_tifffile[lzma-f4] SKIPPED                      [ 99%]
test_imagecodecs.py::test_tifffile[zstd-u1] SKIPPED                      [ 99%]
test_imagecodecs.py::test_tifffile[zstd-u2] SKIPPED                      [ 99%]
test_imagecodecs.py::test_tifffile[zstd-f4] SKIPPED                      [ 99%]
test_imagecodecs.py::test_tifffile[packbits-u1] SKIPPED                  [ 99%]

```

**Comment by cgohlke** [comment]
> Seems like many tests are running, but some seem to be getting skipped

The more [test requirements](https://github.com/cgohlke/imagecodecs/blob/v2020.5.30/imagecodecs/imagecodecs.py#L89-L100) are installed, the less tests are skipped.

If imagecodecs is compiled with all features and all test requirements are installed, there should be about 4563 passed, 1631 xfailed, and no skipped test cases.

**Comment by hmaarrfk** [comment]
got it. 

I guess for now these are the results:
Linux
```
=============== 2088 passed, 3281 skipped, 823 xfailed in 58.70s ===============
```

Mac
```
========== 2087 passed, 3282 skipped, 823 xfailed in 63.62s (0:01:03) ==========
```

Windows
```
========= 2088 passed, 3281 skipped, 823 xfailed in 77.13s (0:01:17) ==========
```

Seems like many are getting skipped for only forgetting a few dependencies. But I guess that is enough for now.

**Comment by cgohlke** [comment]
Yes, looks good to me.

File: recipes/imagecodecs/build.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+# Upstream likes to use crlf, which is really annoying to patch for
+# hmaarrfk
+# Patch included as reference
+cp $RECIPE_DIR/setup.py .
+# Need to add the openjpeg2 cflags
+export CFLAGS="${CFLAGS} $(pkg-config --cflags libopenjp2)"
+
+# Setting LIBRARY_INC is a temporary workaround to enforce the conda-forge code path when building imagecodecs
+# See https://github.com/conda-forge/staged-recipes/pull/10331#issuecomment-636441423
+LIBRARY_INC="bogus" $PYTHON -m pip install . -vv
```
Comment (jakirkham) [inline recipes/imagecodecs/build.sh:11]: Do we still need this workaround now that `setup.py` has been corrected?

```suggestion
$PYTHON -m pip install . -vv
```

File: recipes/imagecodecs/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    # While jpeg-turbo is the preferred version upstream, the migration
+    # to libjpeg-turbo has been contentious on conda-forge
+    # - libjpeg-turbo
+    # Things that are not packaged yet
+    # - brunsli
+    # - zfp
+  run:
+    - python
+    - {{ pin_compatible('numpy') }}
+    - libtiff
```
Comment (jakirkham) [inline recipes/imagecodecs/meta.yaml:53]: Should this be in `host` as well?

File: recipes/imagecodecs/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+test:
+  source_files:
+    - tests
+  requires:
+    - pytest
+  # TODO: add these tests when we start to include builds with turbo
+  # files:
+  #     - 687px-Mona_Lisa,_by_Leonardo_da_Vinci,_from_C2RMF_retouched.jpg
+  #     - ensure_jpegturbo_imagecodecs_plays_nicely_with_jpeg9.py
```
Comment (jakirkham) [inline recipes/imagecodecs/meta.yaml:63]: Is this file in the recipe somewhere?

**Comment by jakirkham** [comment]
Thanks for the work all! Had a couple follow-up questions based on the recent changes.

File: recipes/imagecodecs/meta.yaml (hunk 3 above)
Comment (hmaarrfk) [reply recipes/imagecodecs/meta.yaml:53]: wow thanks for the catch!

File: recipes/imagecodecs/build.sh (hunk 2 above)
Comment (hmaarrfk) [reply recipes/imagecodecs/build.sh:11]: thanks

File: recipes/imagecodecs/meta.yaml (hunk 4 above)
Comment (hmaarrfk) [reply recipes/imagecodecs/meta.yaml:63]: just added it back/

**Comment by jakirkham** [comment]
Thanks @hmaarrfk for all of the work here and everyone for the reviews! Let's pick up anything else in the feedstock. Thanks again all 😄

**Comment by csachs** [comment]
Thank you all for pushing this forward :smile:

**Pull request merged**